### PR TITLE
feat(mono): allow for code definitions to be split

### PR DESCRIPTION
**Pull request by AvarianKnight**
### Goal of this PR
It should be noted the testing for this was very minimal (only tested with one native having `cs_split`)

Allow for C# code gen to do method overloading on `cs_split` parameters

This should let us update to update the native docs argument types without having to worry about C# compatibility, as long as the arguments are properly split.

An example would be if given this native
```c
void NETWORK_FADE_IN_ENTITY(Entity entity, BOOL bNetwork, cs_split BOOL flash);
```

This will generate

```cs
public static NetworkFadeInEntity(int entity, bool bNetwork)
```
and
```cs
public static NetworkFadeInEntity(int entity, bool bNetwork, bool flash)
```

### How is this PR achieving the goal
Adding `cs_split` as a valid doc annotation and splitting the methods when there's a split

### This PR applies to the following area(s)

FiveM, RedM, ScRT: C#

### Successfully tested on
**Game builds:** .. 

**Platforms:** Windows, Linux

### Checklist

- [x] Code compiles and has been tested successfully.
- [x] Code explains itself well and/or is documented.
- [x] My commit message explains what the changes do and what they are for.
- [x] No extra compilation warnings are added by these changes.

### CODE REVIEW

**Comment by Nobelium-cfx** [comment]
Left some style related comments. Overall the change looks reasonable.

But I would like to test and understand it a bit more before we merge it. Im happy to try some tests myself but Im not familiar with how this part of the system works. Can you give me more details on how exactly you use it / where you update the native docs? Do I understand correctly, that after this change we can add new `.md` file `native-decls` with `void NETWORK_FADE_IN_ENTITY(Entity entity, BOOL bNetwork, cs_split BOOL flash);`?

**Review by Nobelium-cfx — CHANGES_REQUESTED** [review]

**Comment by AvarianKnight** [comment]
> Im happy to try some tests myself but Im not familiar with how this part of the system works. Can you give me more details on how exactly you use it / where you update the native docs?

```sh
# goto /ext/native-doc-tooling directory as this is where native gen is done
cd ext/native-doc-tooling
# Apply the patch from this PR https://github.com/citizenfx/native-doc-tooling/pull/19
curl https://github.com/citizenfx/native-doc-tooling/commit/d0938db086e534e6ceb690f392a509ae3e982730.patch | git apply -v
# You will need to clone the natives, we'll just do it into C:\Natives
git clone https://github.com/citizenfx/natives C:\Natives
# we need to use the provided node version or else this wont work
# this will output a `natives_test.json` that we will use for code gen
..\native-doc-gen\node.exe index.js "C:\Natives"
# generate the Lua file that we'll use for codegen, we'll put it into the natives directory for faster usage
..\native-doc-gen\node.exe .\build-template.js >> ../natives/test_native_gen.lua
# goto the native directory so we can build codegen
cd ..\natives
# build native gen and put it into the output folder
.\lua53.exe codegen.lua test_native_gen.lua cs >> out/TestNativeGenOut.cs
```

> Do I understand correctly, that after this change we can add new `.md` file `native-decls` with `void NETWORK_FADE_IN_ENTITY(Entity entity, BOOL bNetwork, cs_split BOOL flash);`?

This would be meant for existing native declarations inside of the [natives](https://github.com/citizenfx/natives) repo.

Certain natives (the example used here is [NetworkFadeInEntity](https://github.com/citizenfx/natives/blob/master/NETWORK/NetworkFadeInEntity.md) have more parameters then are listed in the native docs, but cannot be updated because it would break C# ABI compatibility.

This is meant to allow us to add parameters while maintaining ABI compatibility with C#, we just overload the method.

**Comment by ivanzaida** [comment]
It seems like an unfinished feature since it only covers the C# natives gen part but completely ignores the TS typings generation

**Comment by AvarianKnight** [comment]
> It seems like an unfinished feature since it only covers the C# natives gen part but completely ignores the TS typings generation

There isn't a need to do this for TS type generation, the only breakage there would be that new code needs to update its arguments, old code would still be fine.

**Comment by Nobelium-cfx** [comment]
I've played around and run some tests on these changes - seems to work well for me! Happy to approve after the requested changes above are fixed.

General question: it seems that `codegen_out_cs_v2.lua` solves this issue already. But I assume we don't use it yet? Do you know whats the status there?

File: ext/natives/codegen_out_cs.lua
Comment (Nobelium-cfx) [reply ext/natives/codegen_out_cs.lua:515]: I think you can just revert the `local nativeName = printFunctionName(native) .. (is_split and '' or baseAppendix)` line. And then this change `nativeName .. baseAppendix` shouldn't be needed as well. I think `formatImpl` should work out of the box for you.

Please update the commit when you have time

File: ext/natives/codegen_out_cs.lua
Comment (Nobelium-cfx) [reply ext/natives/codegen_out_cs.lua:7]: I see that this and other Nits were closed without any changes to the code. Is there any reasons you disagree with them? They are minor, but I think it would be good to fix before merging.

**Comment by AvarianKnight** [comment]
> General question: it seems that `codegen_out_cs_v2.lua` solves this issue already. But I assume we don't use it yet? Do you know whats the status there?

The v2 variant is for mono rt2 which is [date locked](https://github.com/citizenfx/fivem/blob/9c623375d87bfc94d2fccef913b8c895c35406c0/code/components/citizen-scripting-mono-v2/src/MonoScriptRuntime.cpp#L267) and will periodically be unusable because of that, but it can change the types without worrying about compatibility because the native invokes are included in the resource.

The regular mono rt is guaranteed to be stable and can't have any ABI breaking changes applied to it so we have to do workarounds to make sure we don't break anything (cs_type(OldType) and now cs_split)

File: ext/natives/codegen_out_cs.lua
Comment (Nobelium-cfx) [inline ext/natives/codegen_out_cs.lua:368]: Nit: redundant space added.

File: ext/natives/codegen_out_cs.lua
Comment (Nobelium-cfx) [inline ext/natives/codegen_out_cs.lua:240]: Nit: this comment is now redundant.

**Review by Nobelium-cfx — APPROVED** [review]
A few more nits. But overall LGTM!

**Comment by Nobelium-cfx** [comment]
Should be merged after: https://github.com/citizenfx/native-doc-tooling/pull/19

**Comment by Nobelium-cfx** [comment]
When merging, also merge https://github.com/citizenfx/native-doc-tooling/pull/19

**Pull request merged**
